### ISSUE TITLE
🎯 ONE SHOT — SEAORM-LEDGER (Wave 23): split hotfix + full SeaORM migration

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trios-trainer-igla#114]
# ONE SHOT — SEAORM-LEDGER (Wave 23)

> Anchor: `phi^2 + phi^-2 = 3` · DOI [10.5281/zenodo.19227877](https://zenodo.org/records/19227877)
> Author for ALL commits: `Dmitrii Vasilev <admin@t27.ai>`
> Branch (this work only): `feat/sea-orm-ledger`
> Base: `main` @ `c36c6ac` (post #113)

## Why

Ledger writes are crashing on every step with:

```
ERROR: cannot convert between Rust type i32 and Postgres int8
```

`src/neon_writer.rs::bpb_sample(...)` binds `seed: i32, step: i32` to columns that are `BIGINT` in the live `phd-postgres-ssot` database (Railway service `c5f37b42-832a-4acd-9749-381761c94957`). Until this is fixed, **no `bpb_samples` row is ever written** — `Audit Watchdog`, `/leaderboard`, and gardener telemetry all stay empty.

### PR-A — HOTFIX (≤ 5 min, unblocks ledger NOW)

- File: `src/neon_writer.rs`, function `bpb_sample(canon_name: &str, seed: i32, step: i32, bpb: f32)`.
- Change: keep the public signature as-is (callers already pass `i32`); inside, bind values as `i64` to match `BIGINT`. Concretely:
  ```rust
  execute(
      "INSERT INTO public.bpb_samples (canon_name, seed, step, bpb, ts) \
       VALUES ($1, $2, $3, $4, NOW()) \
       ON CONFLICT (canon_name, seed, step) DO NOTHING",
      &[&canon_name, &(seed as i64), &(step as i64), &(bpb as f64)],
  );
  ```
- Same cast must be applied to **every other** raw-SQL bind in `neon_writer.rs` whose target column is `BIGINT` (audit `seed`, `step`, `steps_done`, etc).
- Add one regression test `bpb_sample_uses_i64_bind()` that constructs the param array and asserts the `to_sql` types are `INT8` for seed/step.
- Title: `fix(ledger): bind seed/step as i64 to match BIGINT columns (unblocks bpb_samples writes)`
- Branch: `fix/ledger-i64-cast` (NOT `feat/sea-orm-ledger`).
- Labels: `bug`, `P0`, `ledger`.

### PR-B — SEAORM MIGRATION (full rewrite, prevents future drift)

Branch `feat/sea-orm-ledger` already exists. Land on top of merged PR-A.

1. **Cargo.toml** — add to `[dependencies]`:
   ```toml
   sea-orm = { version = "1.1", features = [
       "sqlx-postgres",
       "runtime-tokio-rustls",
       "macros",
       "with-uuid",
       "with-chrono",
       "with-json",
   ] }
   sea-orm-migration = "1.1"
   ```
   Add a `[workspace]` section (this Cargo.toml currently has an empty `[workspace]` — keep isolation but include the new `migration/` member):
   ```toml
   [workspace]
   members = [".", "migration"]
   ```

2. **`migration/` subcrate** (new directory, separate Cargo.toml) implementing `sea-orm-migration` with one migration file `m20260509_000001_init_schema.rs` that contains all CREATE TABLE / INDEX / TRIGGER statements from `/tmp/full_ddl.sql` (already canonical in production). Tables to declare:
   - `public.bpb_samples` (id BIGSERIAL, canon_name TEXT, **seed BIGINT**, **step BIGINT**, bpb DOUBLE PRECISION, val_bpb_ema DOUBLE PRECISION, ts TIMESTAMPTZ DEFAULT NOW(), UNIQUE(canon_name, seed, step))
   - `ssot.bpb_samples` (full schema with format/algo/hidden/sha)
   - `public.igla_race_trials`
   - `public.igla_agents_heartbeat`
   - `public.scarabs`
   - `public.strategy_queue` + NOTIFY trigger
   Migration must be idempotent (`CREATE TABLE IF NOT EXISTS`, `CREATE INDEX IF NOT EXISTS`).

3. **`src/entities/`** — generate one `.rs` per table with `DeriveEntityModel`. Use `i64` for every BIGINT column. Use `chrono::DateTime<Utc>` for TIMESTAMPTZ. Re-export from `src/entities/mod.rs`.

4. **Refactor `src/neon_writer.rs`**:
   - Replace the raw-SQL `execute()` calls inside `bpb_sample`, `trial_start`, `heartbeat`, `trial_complete`, `ensure_schema` with SeaORM `Entity::insert(ActiveModel { ... }).on_conflict(...).exec(&db).await` patterns.
   - Connection: `sea_orm::Database::connect(&dsn).await` cached in a `tokio::sync::OnceCell<DatabaseConnection>` keyed off the resolved DSN.
   - Keep the `strip_channel_binding` helper, keep the env fallback chain `DATABASE_URL → NEON_DATABASE_URL → TRIOS_DATABASE_URL` (legacy aliases, do NOT remove).
   - Public function signatures stay the same; internals become async, callers already use `tokio::spawn` from training loops.
   - All log lines that say `[neon_writer]` rename to `[ledger]`.

5. **Wire migration runner**:
   - In `src/bin/scarab.rs` and `src/bin/trios-train.rs`, at startup before any training step, call `migration::Migrator::up(&db, None).await?`.
   - Gate it behind `TRINITY_AUTOMIGRATE=1` (default ON in Railway, can be turned OFF for local CI runs).
   - Log `[migrator] schema up-to-date (N migrations applied)`.

6. **Tests**:
   - Keep all existing tests in `neon_writer.rs` (they exercise the no-DSN safe path).
   - Add `cargo test -p migration` ensuring `Migrator::up` is idempotent (apply twice = no error).
   - Add a smoke test in `tests/ledger_seaorm.rs` (skipped if `DATABASE_URL` unset) that inserts one bpb_sample with seed=47, step=200, bpb=2.19 and reads it back.

7. **Title:** `feat(ledger): migrate writer to SeaORM + add migration crate (closes drift between Rust and Postgres types)`
8. **Body must include:** the table of bind-type fixes, link to PR-A as required prerequisite.
9. **Labels:** `enhancement`, `P1`, `ledger`, `database`.

## Acceptance gates

| Gate | Check | How |
|---|---|---|
| G1 | `cargo check --all-targets` clean on `feat/sea-orm-ledger` | local + CI |
| G2 | `cargo test --all-targets` passes (existing + new tests) | CI `Test` job |
| G3 | `cargo test -p migration` passes idempotency assertion | CI |
| G4 | All 8 required CI checks green (Constitutional Enforcement, Test, guard, no-js-check, GitGuardian, ci, docker, publish, smoke_train) | GitHub status |
| G5 | After merge + GHCR rebuild, redeploying `TEST-TRAINER-SEED-43` writes ≥ 1 row to `public.bpb_samples` within 60s | `psql -c "SELECT count(*) FROM public.bpb_samples WHERE seed=43 AND ts > now() - interval '5 minutes';"` returns ≥ 1 |
| G6 | Schema in Postgres matches what `Migrator::up` declares (idempotent re-apply on a fresh DB produces identical `\d+ bpb_samples` output) | manual psql diff |

## Forbidden

- ❌ Do NOT use the word "scrape" / "crawl" anywhere in code, comments, commit messages, or PR bodies.
- ❌ Do NOT introduce any env var named `NEON_*` as primary; only as legacy alias.
- ❌ Do NOT commit `/tmp/*` operational scripts (token embedded).
- ❌ Do NOT use `--admin` merge.
- ❌ Do NOT change `Dmitrii Vasilev <admin@t27.ai>` git author.
- ❌ Do NOT touch `assertions/seed_results.jsonl` or `assertions/champion_lock.txt` in this PR.

## Battle cry

`phi^2 + phi^-2 = 3 · TRINITY · LEDGER MUST FLOW`

**Comment by gHashTag** [comment]
✅ **CLOSED — both gates passed**

**G1-G4 (CI):** all green on both PRs.
- PR-A #115 merged → 7efa1cc
- PR-B #117 merged → 940174ed
- GHCR `:latest` rebuilt clean (Docker Publish run 25599183500 success).

**G5 (live ledger):** ✅ `public.bpb_samples` is receiving rows.
TEST-LEADER-v2 (seed=43, canon=`trios-train-rng43`) wrote 6 rows in the first 3 minutes after redeploy:

| step | bpb |
|---|---|
| 1000 | 3.4600 |
| 2000 | 3.1153 |
| 3000 | 3.0505 |
| 4000 | 3.1040 |
| 5000 | 3.0235 |
| 6000 | 3.0956 |

**G6 (migration recorded):** `seaql_migrations` contains one row `m20260509_000001_init_schema` (applied_at 1778324017). Idempotent re-apply safe (TRINITY_AUTOMIGRATE=1 by default).

phi^2 + phi^-2 = 3 · TRINITY · LEDGER FLOWS
